README.md

```diff
@@ -1,2 +1,285 @@
+by removing it from the list.
+
+```yaml
+
+  - name: ladybug-operator
+    inputs:
+      parameters:
+        - name: script
+    container:
+      name: ladybug
+      image: ladybugtools/ladybug:latest
+      command:
+        - python
+      args:
+        - '{{ inputs.parameters.script }}'
+    requirements:
+      pip:
+        - name: lbt-ladybug
+          version: ">=0.4.3"
+      language:
+        - name: python
+          version: ">=3.6"
+      platform:
+       - Windows
+       - Linux
+       - Mac
+
+```
+
+You can also import the operator from a `YAML` or `JSON` file using the `import_from`
+key.
+
+```yaml
+
+  - name: ladybug-operator
+    inputs:
+      parameters:
+        - name: script
+    import_from: /operators/ladybug_operator.yaml
+
+```
+
+## 3. templates
+
+Templates are a discrete reusable unit of code that can be executed separately or as part
+of a workflow. Queenbee supports 4 type of objects as templates:
+
+- `module` : A single module. Each module uses an operator from operators section.
+- `dag`: Collection of modules as a Directed Acyclic Graph (DAG).
+- `steps`: Collection of modules to be executed sequentially in steps.
+- `workflow`: A full Queenbee workflow itself can be referenced as a template in another
+  workflow.
+
+Templates can use the `import_from` keyword to import the `template` from a `YAML` or
+`JSON` file. Keep in mind that a template in a file should be self-sufficient and include
+all the information including `operator`s. This is different from when the `template` is
+part of the `workflow` where an `operator` will be referenced by name.
+
+## 4. flow
+
+`flow` is the heart of the workflow to define how the templates should be executed.
+Queenbee supports two different types of workflows. DAGFlow and SteppedFlow. 
+
+It's common to use the output of one step as an input for another step or reference one
+of the workflow inputs as an input for one of the steps or tasks. Queenbee supports below
+words as prefix variable name:
+
+- workflow: "{{workflow.xx.yy}} is used for workflow level parameters.
+- steps: "{{steps.step_name.xx.yy}} is used in chained templates.
+- tasks: "{{tasks.task_name.xx.yy}} is used in DAG task to refer to other tasks.
+- inputs: "{{inputs.xx.yy}}" is used in operators.
+- item: "{{item}}" or "{{item.key_name}}" is used in loops. You can change item to a
+  different keyword by setting up `loop_var` in `loop_control`.
+
+Here is a example of creating the sky and generating and octree as two consecutive steps.
+Note that the values are place-holders and can be overwritten by input parameters file.
+
+```yaml
+
+flow:
+  - steps:
+    - name: generate_sky_step
+      template: generate_sky
+      inputs:
+        parameters:
+          - name: desired_illuminance
+            value: 100000
+    - name: generate_octree_step
+      template: generate_octree
+      inputs:
+        parameters:
+           - name: scene_files
```

Comment:
How do we differentiate between parameters that are arrays and parameters that are single line strings? Can a parameter also be a JSON?

Reply:
Yes. It can be from any types including JSON. The template should take care of handling different types.

```python
class Parameter(BaseModel):
    """Parameter.

    Parameter indicate a passed string parameter to a service template with an optional
    default value.
    """
    name: str = Schema(
        ...,
        description='Name is the parameter name.'
    )

    value: Any = Schema(
        None,
        description='Default value to use for an input parameter if a value was not'
        ' supplied.'
    )
```

At one point I considered handling them in workflow by adding a `*List` type but then I felt it is out of the scope of a task / step and should be handled in each template itself.

```python
class ArtifactList(Artifact):
    """Artifact List indicates a list of artifact locations."""
    path: List[str] = Schema(
        None,
        description='Path is the path to the artifact which can be a local path or a url.'
    )

    join: str = Schema(
        None,
        description='A string to join inputs together into a single string.'
    )
```

Reply:
Have you tried this with Argo? I'm not sure argo can take non string inputs.

Reply:
I have not. based on the examples it should be able to handle other types of values too. I will double check. If that was the case then we have to use the `join` option to join them together as a string but I think it should not be necessary.